internal/tlsdialer/tlsdialer.go

```diff
@@ -0,0 +1,199 @@
+	tlsconn, state, err := h.Handshaker.Handshake(ctx, conn, config)
+	err = errwrapper.SafeErrWrapperBuilder{
+		Error:     err,
+		Operation: "tls_handshake",
+	}.MaybeBuild()
+	return tlsconn, state, err
+}
+
+// EventsSaver is an Handshaker that saves events
+type EventsSaver struct {
+	Handshaker
+	events []Events
+	mu     sync.Mutex
+}
+
+// ReadEvents reads the saved events and returns them
+func (h *EventsSaver) ReadEvents() []Events {
+	h.mu.Lock()
+	ev := h.events
+	h.events = nil
+	h.mu.Unlock()
+	return ev
+}
+
+// Handshake implements TLSHandshaker.Handshake
+func (h *EventsSaver) Handshake(
+	ctx context.Context, conn net.Conn, config *tls.Config,
+) (net.Conn, tls.ConnectionState, error) {
+	ev := Events{Config: config.Clone(), StartTime: time.Now()}
+	tlsconn, state, err := h.Handshaker.Handshake(ctx, conn, config)
+	ev.EndTime = time.Now()
+	ev.ConnState = state
+	ev.Error = err
+	h.mu.Lock()
+	h.events = append(h.events, ev)
+	h.mu.Unlock()
+	return tlsconn, state, err
+}
+
+// Events contains TLS handshake events
+type Events struct {
+	Config    *tls.Config
+	ConnState tls.ConnectionState
+	EndTime   time.Time
+	Error     error
+	StartTime time.Time
+}
+
+// CleartextDialer is the dialer interface assumed by this package
+type CleartextDialer interface {
+	// DialContext is like net.Dialer.DialContext. It should split the
+	// provided address using net.SplitHostPort, to get a domain name to
+	// resolve. It should use some resolving functionality to map such
+	// domain name to a list of IP addresses. It should then attempt to
+	// dial each of them until one returns success or they all fail.
+	DialContext(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+// Dialer is a dialer that dials TLS connections. This interface is
+// intentionally compatible with the needs of http.Transport.
+type Dialer interface {
+	// DialTLSContext is like Dialer.DialContext except that it will also
+	// perform a TLS handshake using the host part of address.
+	DialTLSContext(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+// StdlibDialer is a TLS dialer using the standard library
+type StdlibDialer struct {
+	CleartextDialer CleartextDialer
+	Handshaker      Handshaker
+}
+
+// DialContext implements Dialer.DialContext
+func (d StdlibDialer) DialContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	return d.DialTLSContext(ctx, network, address)
+}
+
+// DialTLSContext implements TLSDialer.DialTLSContext.
+func (d StdlibDialer) DialTLSContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	return d.DialTLSContextWithConfig(ctx, network, address, nil)
+}
+
+// DialTLSContextWithConfig dials a TLS connection using the specified config, which
+// may be nil, in which case we will use a default config. If config.NextProtos is not
+// specified, then we'll advertise as next protocols "h2" and "http/1.1".
+func (d StdlibDialer) DialTLSContextWithConfig(
```

Comment:
Should not be public, we can change the config on the fly with the handshaker, right?